mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java

```diff
@@ -324,6 +325,16 @@ public final int hashCode() {
      */
     Publisher<MessageId> search(MultimailboxesSearchQuery expression, MailboxSession session, long limit) throws MailboxException;
 
+    /**
+     *
+     * @param threadId
+     *          target Thread
+     * @param session
+     *          the context for this call, not null
+     * @return  a list of MessageId of messages belong to that Thread
+     */
+    Flux<MessageId> getThread(ThreadId threadId, MailboxSession session) throws MailboxException;
```

Comment:
Use Publisher instead of Mono / Flux in TLD API

This allows subcomponents to not be dependant of the reactive technology...

Reply:
No tests for this method?

Reply:
IMO we are testing ThreadIdGuessingAlgorithm in an inconvenient way. We depend on guice to resolve the loop of SearchThreadIdGuessingAlgorithm and MailboxManager, so we can test this from HTTP level. But we are having problems when testing ThreadIdGuessingAlgorithm APIs from the lower level (testing resource based on loop constructors). 

My proposal solutions:
- In the lower level tests: Using MessageMapper to save our custom messages with custom threadId (But we missed some actual guessing threadId)
OR
- SearchThreadIdGuessingAlgorithm depends on MailboxManger no more.

Personally, I would like to do something like inbox.appendMessage, and then getThread directly.
What do you think?

Reply:
Or...

You can implement MailboxManager  by doing directly the search without calling the guessing algorithm?

Reply:
At the moment, we just use searching for InMemory, we still need NaiveThreadIdGuessingAlgorithm for Cassandra and OpenJPA. And I would like to package related ThreadId APIs in ThreadIdGuessingAlgorithm.

What about recoding MailboxManager::search in SearchThreadIdGuessingAlgorithm? (a little duplicated though, but solve above problem)

Reply:
No code duplication is not a solution.